### ISSUE TITLE
250704 Machine Learning Basic_장윤정

### ISSUE THREAD

**Issue by yjjang06** [AIBootcamp14/wiki-rider#11]
## 📅 학습 날짜  
2025-07-04

## 📌 키워드 및 요약 정리
키워드 : 모델 평가, 정성평가, 정량평가, 데이터셋 분할, 모델의 최적화, 과적합

### 🔶 1. 데이터셋 분할

전체 데이터셋을 일정 비율로 학습용 / 검증용 / 평가용 데이터셋으로 분할하여 학습 및 검증, 평가하는데 사용한다.

### 🔹 데이터셋 분할 방법

**평가용 데이터셋**은 모델의 최종 성능을 평가할 때 한 번만 사용하는게 원칙  
**검증용 데이터셋**의 성능 평가 결과는 모델의 구조를 결정하는 하이퍼 파라미터의 최적화, 학습 완료된 모델의 선별에 사용된다.

- **2분할**
  - 전체 데이터셋을 2개로 분할해 학습용(train)과 검증용(validation)으로 사용하는 방법
  - 데이터셋의 양이 적거나 별도의 평가용 데이터셋이 필요하지 않은 경우 사용
- **3분할**
  - 전체 데이터셋을 3개로 분할해 학습용(train), 검증용(validation), 평가용(test)으로 사용하는 방법
  - 일반적으로 3분할 방법을 사용
- **k-fold 교차검증 (Cross Validation)**
  - 데이터셋을 k개의 같은 크기의 부분집합 fold로 나누고, 각 fold를 번갈아가며 학습/검증용 데이터로 사용하는 방법
- **계층적 분할 (Stratified Splitting)**
  - 각 데이터셋을 클래스별 비율을 유지한 채 클래스에서 랜덤 샘플링하여 데이터셋을 구성하는 방법
  - 특정 클래스의 샘플이 데이터셋에 포함되지 않은 경우, 해당 데이터셋은 전체 문제 공간을 충분히 반영하지 못하므로 **대표성**이 떨어질 수 있기 때문에 이 방법을 사용
- **시계열 데이터 분할**
  - 가장 최근의 데이터를 **검증용**, 그보다 이전의 데이터를 **학습용**, 현 시점 이후 미래의 데이터를 **평가용**으로 사용

### 🔹 일반화 성능 (Generalization Ability)
외삽 성능, 내삽 성능을 포함하여 모델이 처음 접하는 데이터에서 얼마나 좋은 성능을 내는지 나타내는 개념

- **내삽(Interpolation)**
  - 학습 데이터셋의 범위(분포) 내에서 새로운 데이터를 예측하거나 분류하는 것
- **외삽(Extrapolation)**
  - 학습 데이터셋의 범위(분포)를 넘어서는 새로운 데이터를 예측하거나 분류하는 것
  - 학습 데이터에 없는 영역에서 패턴을 추정해야 하기 때문에 성능이 낮아지기 쉽다.

> **검증용 데이터셋**은 내삽 범위 성능의 측정을 위해 학습용 데이터셋과 같은 분포를 유지,  
> **평가용 데이터셋**은 외삽 범위 성능까지 측정할 수 있도록, 최대한 다양한 샘플들을 포함해서 구성한다.

---

### 🔶 2. 정량평가
모델의 성능이 얼마나 좋은지 수치로 나타내는 평가 지표(Evaluation Metric) 함수를 설정해두고 이 지표를 활용해 모델 성능을 판단하는 방법. _예) 정확도_

상황에 맞게 여러 정량 평가 지표를 활용하여 평가하며, 여러 개의 정량 평가 지표가 서로 충돌하는 상황에서는  
하나 이상의 지표를 충족 지표로 설정하여 해당 기준을 만족하는지 확인하고, 주요 목표 지표는 최적화 지표로 설정하여 평가한다.

### 🔹 혼동 행렬 (Confusion Matrix)
- 이진 분류 문제에서 모델의 예측치와 실제값에 따른 샘플의 수를 표 형태로 나타낸 것
- 예측이 맞았는지(True) 틀렸는지(False), 결과가 양성(Positive)인지 음성(Negative)
    |  | **예측: Positive (검사 양성)** | **예측: Negative (검사 음성)** |
    | --- | --- | --- |
    | **실제: YES (실제 걸림)** | TP (True Positive) | FN (False Negative) |
    | **실제: NO (걸리지 않음)** | FP (False Positive) | TN (True Negative) |
    - **정밀도 (Precision)** : 모델이 양성으로 예측한 샘플 중 정답(실제 양성)의 비율, TP / (TP + FP)
    - **재현율 (Recall)** : 실제로 양성인 샘플 중 모델이 양성으로 정확하게 분류한 샘플의 비율, TP / (TP + FN)
    - **F1 스코어** : Recall 과 Precision 을 동시에 비교하기 위해 조화평균(Harmonic Mean)을 취한 것

### 🔹 오차
- 예측값 - 실제값
- **MAE(Mean Absolute Error)** 지표는 오차의 크기를 선형적으로 측정, 전체적인 예측 정확도(평균적인 오차)에 더 중점을 둔다
- **MSE(Mean Squared Error)** 지표는 이상치(Outlier)의 존재 여부에 민감하게 반응한다.
#### MAE (Mean Absolute Error)
- 오차에 절대값을 취한 것의 평균
#### MSE (Mean Squared Error)
- 오차를 제곱한 것의 평균
- 결과값의 단위는 원래 데이터 단위의 제곱
#### RMSE (Root Mean Squared Error)
- MSE에 제곱근을 취한 값
- MSE는 단위가 원래 데이터 단위의 제곱된 상태이므로, 이를 보완하기 위해 제곱근을 취해 원래 데이터와 동일한 단위로 복원

---

### 🔶 3. 정성평가
모델에 들어가는 입력값과 그에 따른 모델의 출력값을 사람이 직접 확인해보면서 모델의 성능을 판단하는 방법

정량 평가만으로 모델을 평가할 수 없는 경우가 있기 때문에 정성평가도 필요하다.

### 🔹 지속 가능한 정성평가를 위한 팁
- 풀려는 문제와 모델, 데이터에 대해 잘 알고 있는 한 명 또는 소수의 몇 명이 여러 방법으로 소규모 정성평가를 직접 진행해보며 효율성을 분석한다.
  - 정성평가의 속도 확인, 평가과정에서 비효율적인 부분이 있는지, 개선해볼 방법이 있는지 찾아본다
- 샘플 관찰 계획 세우기
  - 모든 샘플을 한번에 확인하기보다는 우선 일부만 랜덤 샘플링해서 소규모로 진행해본 뒤 추가 진행 여부를 판단한다.
- 점진적인 정성평가 프로세스
  - 틀린 케이스 중 상당수가 한 가지 문제로 인해 발생하는 경우가 많기 때문에 소규모 분석을 통해 간단히 해결 가능한 버그를 걸러낼 수 있다.
  - 학습 데이터셋에도 같은 처리를 적용해 더 높은 성능의 모델로 개선할 수 있다면, 새 모델을 이용해 틀린 샘플을 다시 추려서 진행해보는게 훨씬 효율적이다.
- 발견된 이슈의 기록 및 분류
  - 정성평가를 진행하면서 발견된 문제를 최대한 구체적이고 객관적으로 기록한다
  - 자주 등장하는 이슈들은 간결한 표현으로 카테고리화하여 기록해둔다. (각 이슈가 발생하는 빈도를 확인하고 이를 기반으로 해결 우선순위를 설정)
  - 트리거 샘플들을 모아 별도의 평가용 데이터셋을 구성하면 정량평가만으로도 해당 이슈의 추적이 가능하다
    - 트리거(trigger): 모델의 성능에 문제를 일으키는 인풋 데이터의 패턴
  - 새로운 데이터셋을 수집, 라벨링 하는 과정에서도 이런 경우를 잘 처리하도록 관리해 데이터의 품질을 높인다

---

### 🔶 4. 머신러닝 모델의 최적화
**분산과 편향은 Trade-off 관계이므로 둘 사이의 적절한 균형점을 찾아야 한다.**
- **편향**
  - 예측값과 목표값 사이 발생하는 에러의 평균적인 크기
  - 학습 데이터셋의 성능
- **분산**
  - 모델의 예측값이 서로 퍼져있는 정도
  - 학습/검증 데이터셋의 성능 차이

### 🔹 모델의 복잡도, 용량 (Complexity, Capacity)
- 모델이 더 복잡하고 어려운 데이터까지 학습할 수 있는 용량
- 파라미터의 수(딥러닝의 경우 레이어 및 뉴런 수)가 많을수록 복잡도가 커진다.
- 모델의 복잡도가 커질수록 학습을 위해 다양하고 고품질의 데이터가 필요하다.

### 🔹 과적합 (Overfitting)
- 모델이 학습데이터의 패턴에만 과하게 학습되어 학습 데이터에서의 손실함수 값은 0에 수렴하지만 평가 데이터셋 등 새로운 데이터에 대해서는 성능을 내지 못하는 상황
- 큰 분산(Variance) 때문에 발생
- 모델의 복잡도 > 학습 데이터의 복잡성 ➡️ 모델 복잡도를 줄이거나 학습 데이터를 늘린다.

### 🔹 과소적합 (Underfitting)
- 모델이 학습 데이터셋을 충분히 학습하지 못해 최적화가 완료되지 않은 상태
- 모델이 학습 데이터에 대해 낮은 성능을 보이며, 데이터의 구조나 복잡한 패턴을 학습하지 못한 경우
- 큰 편향(Bias) 때문에 발생
- 모델 복잡도 < 학습 데이터의 복잡성 ➡️ 모델 복잡도를 늘리거나 학습 데이터를 줄인다

#### Regularization
- 과적합 문제의 방지를 위해 모델의 복잡도가 필요 이상으로 커지지 않도록 제한을 두는 방법

#### L1, L2 Regularization
- 모델 파라미터(가중치)의 norm(L1 또는 L2)을 손실 함수에 패널티 항으로 더하는 것
  - **L1 (Lasso)** : 모델 파라미터(가중치)의 절대값 합을 패널티로 추가
  - **L2 (Ridge)** : 모델 파라미터(가중치)의 제곱합을 패널티로 추가

#### 학습 데이터의 증강 (Augmentation)
- 정답 라벨이 바뀌지 않는 한도 내에서 인풋 데이터를 여러 방법으로 변형 시켜서 새로운 데이터로 활용하는 방식
- 예 ) 이미지 좌우/상하반전

#### Early Stopping
- 학습, 검증용 데이터셋의 손실함수 값 차이가 커지기 시작하는 시점에 학습을 일찍 종료하여 과적합 문제가 생기지 않도록 하는 방법

---

## ❓ 어려웠던 부분  
- 모델 평가 방식이 다양하다보니 금방 잊어버릴 것 같아 복습을 해야겠다고 생각했습니다.
- 이론 위주로 학습하다 보니, 실제 상황에서 어떻게 평가 지표를 적용해야 할지 감이 잘 오지 않았습니다.

## 💬 질문 또는 토론 유도  
- 실제 모델을 평가할 때 이런 방법들을 어떻게 적용하는지, 평가 지표들은 어떻게 구해서 활용하는지 궁금합니다.  
`scikit-learn`을 사용하여 정량평가 지표를 확인해보고 모델을 평가하는 것일까요?
정량 평가 지표를 보더라도 그 지표를 보고 모델의 성능을 평가하는 것은 사람의 몫일 것 같은데..
어떠한 기준으로 모델을 평가해야 하는지도 궁금합니다.

## 🔗 참고 자료 (선택)  
- [scikit-learn : Model selection and evaluation](https://scikit-learn.org/0.21/model_selection.html)

**Comment by Young-SeokOh** [comment]
윤정 선생님, 안녕하세요 :) 제가 학습 내용을 미처 확인을하지 못해 이제서야 피드백을 드립니다. 미안합니다. 우선 강의내용을 토대로 평가지표에 대해 잘 정리해주셨습니다. 다만, 질문을 주셨듯이, 도메인에 따라 우선하는 지표가 다릅니다. 이 부분은 선생님께서 관심가지고 계신 분야에서 어떤 평가지표를 사용하는지 직접 확인하시는 작업이 필요합니다. 분야마다 데이터가 다르고, 우선하는 지표가 다르기에 강의에서는 이론적으로 접근할 수 밖에 없었을 것입니다.
그리고 최근에는 정성적인 평가도 중요해지고 있습니다. 따라서 해당 부분도 정리를 해두실 필요가 있습니다.

평가지표와 관련해서 정리된 논문을 추천 논문으로 올려드립니다. 평안한 하루되세요 :)

[s41598-024-56706-x.pdf](https://github.com/user-attachments/files/21099982/s41598-024-56706-x.pdf)
